### DISCUSSION TITLE
关于p256.go中从预计算表中查找的问题

### DISCUSSION THREAD

**Discussion by emmansun** [emmansun/gmsm#27] [Q&A]
下面的实现中，每次查找都要把表中15个点扫一遍，这是出于什么目的呢？抗旁路攻击？
```
// p256SelectAffinePoint sets {out_x,out_y} to the index'th entry of table.
// On entry: index < 16, table[0] must be zero.
func p256SelectAffinePoint(xOut, yOut *[p256Limbs]uint32, table []uint32, index uint32) {
	for i := range xOut {
		xOut[i] = 0
	}
	for i := range yOut {
		yOut[i] = 0
	}

	for i := uint32(1); i < 16; i++ {
		mask := i ^ index
		mask |= mask >> 2
		mask |= mask >> 1
		mask &= 1
		mask--
		for j := range xOut {
			xOut[j] |= table[0] & mask
			table = table[1:]
		}
		for j := range yOut {
			yOut[j] |= table[0] & mask
			table = table[1:]
		}
	}
}
```
通过下列代码，我们可以看到mask的值：
```
	for index := uint32(0); index < 16; index++ {
		fmt.Printf("index=%02x, mask=", index)
		for i := uint32(1); i < 16; i++ {
			mask := i ^ index
			mask |= mask >> 2
			mask |= mask >> 1
			mask &= 1
			mask--
			fmt.Printf("%08x ", mask)
		}
		fmt.Println()
	}

结果：
index=00, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=01, mask=ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=02, mask=00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=03, mask=00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=04, mask=00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=05, mask=00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=06, mask=00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=07, mask=00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=08, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 00000000 
index=09, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 00000000 
index=0a, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 00000000 
index=0b, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 00000000 
index=0c, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 00000000 
index=0d, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 00000000 
index=0e, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff 00000000 
index=0f, mask=00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000 ffffffff
```

**Comment by xuyang2** [comment]
标准库代码原提交标题中有提到 constant-time , 似乎是从 chromium, openssl 的代码翻过来的

- https://github.com/golang/go/commit/d2a19e9fd1c4f2c4941d86e860be0cee8c418170#diff-9385ac0a04e65897fc4b40b34c0b5474ea159e2f14a9e50a8c32360931c15aa6R946 ( src/pkg/crypto/elliptic/p256.go Load diff )
- https://golang.org/cl/10551044

<del>另外从 rusttls 项目找到这个 https://github.com/briansmith/ring/blob/main/crypto/fipsmodule/ec/p256.c 
然后就看不懂了</del>

估计要看论文找 Side Channel Attack/Protect 相关的才能确认了
- [Efficient and Secure Elliptic Curve Cryptography Implementation of Curve P-256](https://csrc.nist.gov/csrc/media/events/workshop-on-elliptic-curve-cryptography-standards/documents/papers/session6-adalier-mehmet.pdf)
- [ Fast Elliptic Curve Cryptography in OpenSSL](https://research.google/pubs/pub37376/)
- [Fast Prime Field Elliptic Curve Cryptography with 256 Bit Primes](https://eprint.iacr.org/2013/816.pdf)
- https://www.mail-archive.com/openssl-dev@openssl.org/msg33100.html

**Reply by emmansun** [reply]
应该就是以constant time 执行，达到抗SCA的目的，那篇NIST的论文很好！

**Comment by emmansun** [comment]
It's a **SafeSelect** implementation.  We can reference **4.2 Group Level Optimizations** of the paper "Efficient and Secure Elliptic Curve Cryptography Implementation of Curve P-256".

> **SafeSelect** is an implementation specific function to select the appropriate table entry without leaking information about the secret scalar. Emilia Kasper [36] and Shay Gueron [37] provide well-written examples of performing the SafeSelect function and their code is available as Open Source within the OpenSSL code base [38].
> 
> 36. E. Käsper, “ Fast Elliptic Curve Cryptography in OpenSSL,” Financial Cryptography and Data Security. LNCS, vol. 7126, pp. 27--39. Springer, Heidelberg, 2012.
> 37. S. Gueron, V. Krasnov, “Fast Prime Field Elliptic Curve Cryptography with 256 Bit Primes,” Journal of Cryptographic Engineering, November 2014. https://eprint.iacr.org/2013/816.pdf
> 38. S. Gueron, V. Krasnov, [PATCH] “Fast and side channel protected implementation of the NIST P-256 Elliptic Curve for x86-64 platforms,” OpenSSL patch, October 2013
